pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,364 @@
+
+
+def expand_multi_par(tape, depth=10):
+    """Expand a tape until it does not contain any multi-parameter gates, if possible.
+
+    Args:
+        tape (.QuantumTape): Tape to be expanded
+        depth (int): Maximum expansion depth
+
+    Returns
+        .QuantumTape: Expanded tape
+
+    """
+    if any(len(op.parameters) > 1 for op in tape.operations):
+        stopping_cond = lambda g: (
+            isinstance(g, qml.measure.MeasurementProcess) or len(g.parameters) <= 1
+        )
+        new_tape = tape.expand(depth=depth, stop_at=stopping_cond)
+        params = new_tape.get_parameters(trainable_only=False)
+        new_tape.trainable_params = qml.math.get_trainable_indices(params)
+
+        return new_tape
+
+    return tape
+
+
+def advanced_spectrum(qnode, encoding_args=None, argnum=None, decimals=5, validation_kwargs=None):
+    r"""Compute the frequency spectrum of the Fourier representation of quantum circuits,
+    including classical preprocessing.
+
+    The circuit must only use single-parameter gates of the form :math:`e^{-i x_j G}` as
+    input-encoding gates, which allows the computation of the spectrum by inspecting the gates'
+    generators :math:`G`. The most important example of such gates are Pauli rotations.
+
+    .. note::
+
+        More precisely, the ``advanced_spectrum`` function relies on the gate to
+        define a ``generator``, and will fail if gates marked controlled by marked
+        parameters do not have this attribute.
+
+    The argument ``argnum`` controls which QNode arguments are considered as encoded
+    inputs and the spectrum is computed only for these arguments.
+    The input-encoding *gates* are those that are controlled by input-encoding QNode arguments.
+    If no ``argnum`` are given, all QNode arguments are considered to be input-encoding
+    arguments.
+
+    .. note::
+
+        Arguments or parameters in an argument that do not contribute to the Fourier series
+        of the QNode with a frequency are considered as contributing with a constant term.
+        That is, a parameter that does not control any gate has the spectrum ``[0]``.
+
+    Args:
+        qnode (pennylane.QNode): :class:`~.pennylane.QNode` to compute the spectrum for
+        encoding_args (dict[str, list[tuple]], set): Parameter index dictionary;
+            keys are argument names, values are index tuples for that argument
+            or an ``Ellipsis``. If a ``set``, all values are set to ``Ellipsis``.
+            The contained argument and parameter indices indicate the scalar variables
+            for which the spectrum is computed
+        argnum (list[int]): Numerical indices for arguments with respect to which
+            to compute the spectrum
+        decimals (int): number of decimals to which to round frequencies.
+        num_pos (int): Number of additional random positions at which to evaluate the
+            Jacobian of the preprocessing and test that it is constant.
```

Comment:
The arguments detailed here don't match the function header; `validation_kwargs` is missing.